File: app/src/main/java/tn/esprit/wayfinder/models/BookingModels.kt
package tn.esprit.wayfinder.models

import kotlinx.serialization.Contextual
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class Offer(
    val id: String,
    val type: String,
    val destination: String,
    val price: Double
)

@Serializable
data class OfferComparison(
    @SerialName("offer_id") val offerId: String,
    @SerialName("base_price") val basePrice: Double,
    val taxes: Double,
    val baggage: Double,
    @SerialName("service_fees") val serviceFees: Double,
    val total: Double
)

@Serializable
data class TripDetails(
    val origin: String? = null,
    val destination: String? = null,
    @SerialName("departure_date") val departureDate: String? = null,
    @SerialName("return_date") val returnDate: String? = null,
    @SerialName("travel_class") val travelClass: String? = null,
    val seats: String? = null
)

@Serializable
data class BookingPassenger(
    @SerialName("full_name") val fullName: String,
    @SerialName("traveler_type") val travelerType: String? = null,
    @SerialName("document_number") val documentNumber: String? = null
)

@Serializable
data class Booking(
    @SerialName("_id") val id: String,
    @SerialName("user_id") val userId: String,
    @SerialName("offer_id") val offerId: String,
    val status: BookingStatus,
    @SerialName("payment_details") val paymentDetails: Map<String, @Contextual Any>,
    @SerialName("booking_date") val bookingDate: String,
    @SerialName("confirmation_number") val confirmationNumber: String,
    @SerialName("total_price") val totalPrice: Double,
    @SerialName("trip_details") val tripDetails: TripDetails? = null,
    val passengers: List<BookingPassenger>? = emptyList(),
    val notes: String? = null,
    @SerialName("createdAt") val createdAt: String,
    @SerialName("updatedAt") val updatedAt: String
)

@Serializable
data class CreateBookingRequest(
    @SerialName("offer_id") val offerId: String,
    @SerialName("payment_details") val paymentDetails: Map<String, @Contextual Any>,
    @SerialName("trip_details") val tripDetails: TripDetails? = null,
    val passengers: List<BookingPassenger>? = null,
    val notes: String? = null,
    @SerialName("total_price") val totalPrice: Double? = null
)

@Serializable
data class UpdateBookingRequest(
    @SerialName("payment_details") val paymentDetails: Map<String, @Contextual Any>? = null,
    @SerialName("trip_details") val tripDetails: TripDetails? = null,
    val passengers: List<BookingPassenger>? = null,
    val notes: String? = null,
    val status: BookingStatus? = null,
    @SerialName("total_price") val totalPrice: Double? = null
)

enum class BookingStatus {
    PENDING,
    CONFIRMED,
    CANCELLED
}

File: WayFinder/Models/BookingModels.swift
import Foundation

enum BookingStatus: String, Decodable {
    case pending = "pending"
    case confirmed = "confirmed"
    case cancelled = "cancelled"
    
    var displayName: String {
        switch self {
        case .pending: return "En attente"
        case .confirmed: return "Confirmé"
        case .cancelled: return "Annulé"
        }
    }
}

struct BookingOffer: Decodable, Identifiable {
    let id: String
    let destination: String
    let price: Double
    let type: String? // "flight", "hotel", etc.
    let currency: String? // Optionnel car peut ne pas être dans le JSON
    let departureDate: String?
    let returnDate: String?
    let airline: String?
    let description: String?
    
    enum CodingKeys: String, CodingKey {
        case id
        case destination
        case price
        case type
        case currency
        case departureDate = "departure_date"
        case returnDate = "return_date"
        case airline
        case description
    }
}

struct Booking: Decodable, Identifiable {
    let id: String
    let destination: String
    let status: BookingStatus
    let confirmationNumber: String
    let createdAt: String
    let price: Double?
    let currency: String?
    let departureDate: String?
    let returnDate: String?
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case status
        case confirmationNumber = "confirmation_number"
        case createdAt = "createdAt"
        case price = "total_price"
        case currency
        case offerId = "offer_id"
        case tripDetails = "trip_details"
        case departureDate = "departure_date"
        case returnDate = "return_date"
    }
    
    enum TripDetailsKeys: String, CodingKey {
        case destination
        case departure_date
        case return_date
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        
        // Decode _id to id
        id = try container.decode(String.self, forKey: .id)
        
        // Decode status
        status = try container.decode(BookingStatus.self, forKey: .status)
        
        // Decode confirmation_number
        confirmationNumber = try container.decode(String.self, forKey: .confirmationNumber)
        
        // Decode createdAt - Mongoose timestamps return createdAt
        // Try as String first, then as Date
        let dateFormatter = ISO8601DateFormatter()
        dateFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        
        if let createdAtStr = try? container.decode(String.self, forKey: .createdAt) {
            createdAt = createdAtStr
        } else if let createdAtDate = try? container.decode(Date.self, forKey: .createdAt) {
            createdAt = dateFormatter.string(from: createdAtDate)
        } else {
            // Fallback: use current date as ISO8601 string
            createdAt = dateFormatter.string(from: Date())
        }
        
        // Decode total_price to price
        price = try? container.decode(Double.self, forKey: .price)
        
        // Decode currency (might not exist)
        currency = try? container.decode(String.self, forKey: .currency)
        
        // Decode destination from trip_details.destination or use offer_id as fallback
        var decodedDestination: String? = nil
        var decodedDepartureDate: String? = nil
        var decodedReturnDate: String? = nil
        
        // Try to get destination and dates from trip_details
        if let tripDetails = try? container.nestedContainer(keyedBy: TripDetailsKeys.self, forKey: .tripDetails) {
            decodedDestination = try? tripDetails.decode(String.self, forKey: .destination)
            decodedDepartureDate = try? tripDetails.decode(String.self, forKey: .departure_date)
            decodedReturnDate = try? tripDetails.decode(String.self, forKey: .return_date)
        }
        
        // Fallback to offer_id if destination is not in trip_details
        if decodedDestination == nil || decodedDestination?.isEmpty == true {
            decodedDestination = try? container.decode(String.self, forKey: .offerId)
        }
        
        // Fallback to "N/A" if still no destination
        destination = decodedDestination ?? "N/A"
        
        // If dates weren't found in trip_details, try root level
        if decodedDepartureDate == nil {
            decodedDepartureDate = try? container.decode(String.self, forKey: .departureDate)
        }
        if decodedReturnDate == nil {
            decodedReturnDate = try? container.decode(String.self, forKey: .returnDate)
        }
        
        // Initialize dates once
        departureDate = decodedDepartureDate
        returnDate = decodedReturnDate
    }
    
    // Initializer pour les previews et tests
    init(id: String, destination: String, status: BookingStatus, confirmationNumber: String, createdAt: String, price: Double?, currency: String?, departureDate: String?, returnDate: String?) {
        self.id = id
        self.destination = destination
        self.status = status
        self.confirmationNumber = confirmationNumber
        self.createdAt = createdAt
        self.price = price
        self.currency = currency
        self.departureDate = departureDate
        self.returnDate = returnDate
    }
}

// Note: ConfirmBookingRequest is now handled directly with JSONSerialization in BookingService
// This struct is kept for reference but not used directly
struct ConfirmBookingRequest: Encodable {
    let offerId: String
    let paymentDetails: [String: String]
    let totalPrice: Double?
    
    enum CodingKeys: String, CodingKey {
        case offerId = "offer_id"
        case paymentDetails = "payment_details"
        case totalPrice = "total_price"
    }
}

struct ConfirmBookingResponse: Decodable {
    let id: String
    let confirmationNumber: String
    let status: BookingStatus
    let destination: String?
    let createdAt: String
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case confirmationNumber = "confirmation_number"
        case status
        case destination
        case offerId = "offer_id"
        case createdAt = "createdAt"
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        
        // Decode _id to id
        id = try container.decode(String.self, forKey: .id)
        
        // Decode confirmation_number to confirmationNumber
        confirmationNumber = try container.decode(String.self, forKey: .confirmationNumber)
        
        // Decode status
        status = try container.decode(BookingStatus.self, forKey: .status)
        
        // Try to decode destination, fallback to offer_id if not present
        let decodedDestination = try? container.decode(String.self, forKey: .destination)
        destination = decodedDestination ?? (try? container.decode(String.self, forKey: .offerId))
        
        // Decode createdAt
        createdAt = try container.decode(String.self, forKey: .createdAt)
    }
}

struct TravelDates: Codable {
    let departure: String?
    let returnDate: String?
    
    enum CodingKeys: String, CodingKey {
        case departure
        case returnDate = "return"
    }
}

struct OfferComparison: Decodable {
    let offerId: String
    let alternatives: [BookingOffer]?
    let recommendations: [String]?
    let bestPrice: Double?
    let averagePrice: Double?
    
    enum CodingKeys: String, CodingKey {
        case offerId = "offer_id"
        case alternatives
        case recommendations
        case bestPrice = "best_price"
        case averagePrice = "average_price"
    }
}

struct CreateBookingRequest: Encodable {
    let destinationId: String?
    let offerId: String?
    let travelDates: TravelDates?
    let passengers: Int?
    let totalPrice: Double?
    let currency: String?
    let status: String?
    
    enum CodingKeys: String, CodingKey {
        case destinationId = "destination_id"
        case offerId = "offer_id"
        case travelDates = "travel_dates"
        case passengers
        case totalPrice = "total_price"
        case currency
        case status
    }
}

struct UpdateBookingRequest: Encodable {
    let status: String?
    let travelDates: TravelDates?
    let passengers: Int?
    let totalPrice: Double?
    let currency: String?
    
    enum CodingKeys: String, CodingKey {
        case status
        case travelDates = "travel_dates"
        case passengers
        case totalPrice = "total_price"
        case currency
    }
}

